model/task/task.go

```diff
@@ -3128,6 +3128,102 @@ func GetTaskStatsByVersion(versionID string, opts GetTasksByVersionOptions) ([]*
 	return StatusCount, nil
 }
 
+type GroupedStatusCount struct {
+	Variant      string         `bson:"variant"`
+	DisplayName  string         `bson:"display_name"`
+	StatusCounts []*StatusCount `bson:"status_counts"`
+}
+
+func GetGroupedTaskStatsByVersion(versionID string, opts GetTasksByVersionOptions) ([]*GroupedStatusCount, error) {
+	StatusCount := []*GroupedStatusCount{}
+	pipeline := getTasksByVersionPipeline(versionID, opts)
+	project := bson.M{"$project": bson.M{
+		BuildVariantKey:            "$" + BuildVariantKey,
+		BuildVariantDisplayNameKey: "$" + BuildVariantDisplayNameKey,
+		DisplayStatusKey:           "$" + DisplayStatusKey,
+	}}
+	pipeline = append(pipeline, project)
+	variantStatusesKey := "variant_statuses"
+	groupByStatusPipeline := []bson.M{
+		{
+			"$group": bson.M{
+				"_id": "$" + BuildVariantKey,
+				variantStatusesKey: bson.M{
+					"$push": bson.M{
+						DisplayStatusKey:           "$" + DisplayStatusKey,
+						BuildVariantKey:            "$" + BuildVariantKey,
+						BuildVariantDisplayNameKey: "$" + BuildVariantDisplayNameKey,
+					},
+				},
+			},
+		},
+		{
+			"$unwind": bson.M{
+				"path":                       "$" + variantStatusesKey,
+				"preserveNullAndEmptyArrays": false,
+			},
+		},
+		{
+			"$project": bson.M{
+				variantStatusesKey: 1,
+				"_id":              0,
+			},
+		},
+		{
+			"$group": bson.M{
+				"_id": bson.M{
+					DisplayStatusKey:           "$" + bsonutil.GetDottedKeyName(variantStatusesKey, DisplayStatusKey),
+					BuildVariantKey:            "$" + bsonutil.GetDottedKeyName(variantStatusesKey, BuildVariantKey),
+					BuildVariantDisplayNameKey: "$" + bsonutil.GetDottedKeyName(variantStatusesKey, BuildVariantDisplayNameKey),
+				},
+				"count": bson.M{"$sum": 1},
+			},
+		},
+		{
+			"$sort": bson.M{
+				bsonutil.GetDottedKeyName("_id", DisplayStatusKey): 1,
```

Comment:
Is it necessary that we sort here when we sort at 3207? Or can we combine these at the end to be one step?

Reply:
This sorts the statuses so that they are ordered within the status_count arrays below. Without this step "status_counts" will not be sorted by status.